### PR TITLE
Core: Allow injecting resources in REST and JDBC catalogs

### PR DESCRIPTION
**Pull request by electrum**
Allow catalogs to be stateless by injecting the long-lived resources.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java [hunk 1]
```diff
@@ -127,6 +136,19 @@ public void setFileIOBuilder(Function<Map<String, String>, FileIO> newIOBuilder)
...
   }
 
+  public void setClientPoolBuilder(
+      Function<Map<String, String>, JdbcClientPool> newClientPoolBuilder) {
+    Preconditions.checkState(
+        null == connections, "Cannot set client pool builder after calling initialize");
+    this.clientPoolBuilder = newClientPoolBuilder;
+  }
+
+  public void setInitializeCatalogTables(boolean initializeCatalogTables) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:146]: Why is this needed?

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java [hunk 2]
```diff
@@ -127,6 +136,19 @@ public void setFileIOBuilder(Function<Map<String, String>, FileIO> newIOBuilder)
     this.ioBuilder = newIOBuilder;
   }
 
+  public void setClientPoolBuilder(
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:139]: The IO builder seemed like a reasonable thing to get in quickly, but it does conflict with the `initialize` pattern. If we are going to inject more resources then we should come up with a standard way to do that rather that creating a bunch of these setter methods.

I think what we should probably do is inject these hooks in the constructor. Would that work for you, @electrum?

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 3]
```diff
@@ -126,7 +126,7 @@ public RESTSessionCatalog() {
     this(config -> HTTPClient.builder(config).uri(config.get(CatalogProperties.URI)).build());
   }
 
-  RESTSessionCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
+  public RESTSessionCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:129]: In general, I'm okay with this approach. I think setting these in the constructor is probably the right way to go if we need to do more of this. We should also remove the new setter method if we're opening this up and add the builder here.

I'm also curious why this is needed, since I think that catalogs probably will have their own client and the set of catalogs doesn't really change in Trino. What do you intend to do differently when you can use this?

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java (hunk 1 above)
Comment (electrum) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:146]: Since we create the catalog class for every Trino query, we don’t want to run this repeatedly.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java (hunk 2 above)
Comment (electrum) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:139]: Agreed, constructor injection is better. I was trying to follow the existing pattern.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 3 above)
Comment (electrum) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:129]: The Trino FileIO wrapper is created per Trino query, since it has context about the user (for example, HDFS with Kerberos impersonation). Thus, we need to create the Iceberg catalog per query as well.

The mismatch is the scope of the services. FileIO has query scope, but the HTTP client has singleton scope.

We need to determine the appropriate scope for everything inside the REST catalog and make it injectable.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java (hunk 2 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:139]: Yeah, let's switch over to constructor. I wasn't sure what would work for Trino originally but that's a better pattern.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java (hunk 1 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:146]: Got it. We should just have a `initialized` flag on the constructor, too.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 3 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:129]: Okay, I think we settled this offline. We can move forward with something better for REST since we already handle sessions for REST and don't need a catalog instance per user.

**Comment by electrum** [comment]
Replaced by #7087 and #7088
